src/session.js

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2021 Splunk Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {findCookieValue, generateId, isIframe} from './utils';
+
+/*
+    The basic idea is to let the browser expire cookies for us "naturally" once
+    IntactivityTimeout is reached.  Activity (including any page load) 
+    extends the session.  The true startTime of the session is set in the cookie value
+    and if an extension would ever exceed MaxAge it doesn't happen.
+    We use a background periodic timer to check for expired cookies and initialize new ones.
+    Session state is stored in the cookie as uriencoded json and is of the form
+    { 
+        id: 'sessionIdAsHex',
+        startTime: startTimeAsNewDate_getTime
+    }
+    Future work can add more fields though note that the fact that the value doesn't change 
+    once created makes this very robust when used in multiple tabs/windows - tabs don't compete/
+    race to do anything but set the max-age.
+*/
+
+
+const MaxSessionAgeMillis = 4 * 60 * 60 * 1000;
+const InactivityTimeoutSeconds = 15 * 60;
+const PeriodicCheckSeconds = 60;
+export const COOKIE_NAME = '_splunk_rum_sid';
+
+let rumSessionId = generateId(64); // will be overwritten in init() with scriptInstance; only here as extreme fallback
+let recentActivity = false;
+function markActivity() {
+  recentActivity = true;
```

Comment:
`lastActivityTime = performance.now()` would be more readable, otherwise you have to guess when was the last time it was cleared, which could be difficult if a bug had something to do with inconsistency of `setInverval`

Reply:
Not sure if you meant it this way, but I gather a concern here is the idea that we could mark recent activity, the tab is backgrounded, and an hour later the tab is foregrounded with `updateSessionStatus` running only then (and seeing the "old" `recentActivity`).  Assuming the browser does in fact expire the session for us in the interim, the first thing the periodic task will do is notice the expired cookie and start a new one -> no false continuation of the stale session.  That said, it would be cleaner/timelier to make sure that `updateSessionStatus` also gets called on foregrounding, so I'll add that.